File: code/jvm/im-domain/src/main/kotlin/im/domain/wrappers/password/PasswordValidationError.kt
package im.domain.wrappers.password

sealed class PasswordValidationError(
    private val defaultMessage: String = "Password is invalid",
) {
    data object Blank : PasswordValidationError("Password cannot be blank")

    data class InvalidLength(
        val min: Int,
        val max: Int,
    ) : PasswordValidationError("Password must be between $min and $max characters")

    data class NotEnoughUppercaseLetters(
        val min: Int,
    ) : PasswordValidationError("Password must contain at least $min uppercase letter(s)")

    data class NotEnoughLowercaseLetters(
        val min: Int,
    ) : PasswordValidationError("Password must contain at least $min lowercase letter(s)")

    data class NotEnoughDigits(
        val min: Int,
    ) : PasswordValidationError("Password must contain at least $min digit(s)")

    data object CannotContainWhitespace : PasswordValidationError("Password cannot contain whitespace")

    fun toErrorMessage(): String = defaultMessage
}

fun List<PasswordValidationError>.toErrorMessage(): String = joinToString("\n") { it.toErrorMessage() }

File: code/js/src/Domain/wrappers/password/PasswordValidationError.ts
export abstract class PasswordValidationError {
    private readonly defaultMessage: string;

    protected constructor(defaultMessage: string = 'Password is invalid') {
        this.defaultMessage = defaultMessage;
    }

    static Blank = new (class extends PasswordValidationError {
        constructor() {
            super('Password cannot be blank');
        }
    })();

    static InvalidLength(min: number, max: number) {
        return new (class extends PasswordValidationError {
            constructor() {
                super(`Password must be between ${min} and ${max} characters`);
            }
        })();
    }

    static NotEnoughUppercaseLetters(min: number) {
        return new (class extends PasswordValidationError {
            constructor() {
                super(`Password must contain at least ${min} uppercase letter(s)`);
            }
        })();
    }

    static NotEnoughLowercaseLetters(min: number) {
        return new (class extends PasswordValidationError {
            constructor() {
                super(`Password must contain at least ${min} lowercase letter(s)`);
            }
        })();
    }

    static NotEnoughDigits(min: number) {
        return new (class extends PasswordValidationError {
            constructor() {
                super(`Password must contain at least ${min} digit(s)`);
            }
        })();
    }

    static Insecure = new (class extends PasswordValidationError {
        constructor() {
            super('This password is too insecure, choose a stronger one');
        }
    })();

    static CannotContainWhitespace = new (class extends PasswordValidationError {
        constructor() {
            super('Password cannot contain whitespace');
        }
    })();

    toErrorMessage(): string {
        return this.defaultMessage;
    }
}
